### PR TITLE
Added TryParse to StandardFormat

### PR DESCRIPTION
**Pull request by AlexejLiebenthal**
Fixes issue #32950 

1. Removed StandardFormat.Parse(string) overload in src/System.Memory/ref/System.Memory.cs
2. Added StandardFormat.TryParse in src/System.Memory/ref/System.Memory.cs
3. Marked Obsolete StandardFormat.Parse(string) overload in src/Common/src/CoreLib/System/Buffers/StandardFormat.cs
4. Added TryParse in src/Common/src/CoreLib/System/Buffers/StandardFormat.cs
5. Added UnitTests for TryParse in src/System.Memory/tests/ParsersAndFormatters/StandardFormatTests.cs

Need help for this. I didn't get it exactly how all these parts work together.
Like I see, there is [src/System.Memory/ref/System.Memory.cs](https://github.com/dotnet/corefx/blob/master/src/System.Memory/ref/System.Memory.cs), where the "abstraction" is defined and the [src/Common/src/CoreLib/System/Buffers/StandardFormat.cs](https://github.com/dotnet/corefx/blob/master/src/Common/src/CoreLib/System/Buffers/StandardFormat.cs), but this file has no *.sln or *.csproj. But as you can see regardless of the changes I made the build fails.

### CODE REVIEW

File: src/Common/src/CoreLib/System/Buffers/StandardFormat.cs [hunk 1]
```diff
@@ -71,8 +71,30 @@ public StandardFormat(char symbol, byte precision = NoPrecision)
...
         {
+            ParseHelper(format, out StandardFormat standardFormat, true);
+
+            return standardFormat;
+        }
+
+        /// <summary>
+        /// Converts a classic .NET format string into a StandardFormat
+        /// </summary>
+        [System.Obsolete]public static StandardFormat Parse(string format) => format == null ? default : Parse(format.AsSpan());
```
Comment (stephentoub) [inline src/Common/src/CoreLib/System/Buffers/StandardFormat.cs]: Why is this being marked as Obsolete?  I don't see that as having been part of the approved changes, and we generally do not do that.

**Comment by stephentoub** [comment]
cc: @ahsonkhan

File: src/Common/src/CoreLib/System/Buffers/StandardFormat.cs (hunk 1 above)
Comment (AlexejLiebenthal) [reply src/Common/src/CoreLib/System/Buffers/StandardFormat.cs]: I see. Will change it back.

As I mentioned [here](https://github.com/dotnet/corefx/issues/32950#issuecomment-441552233), I also think now, that the string overloads could still be quite important for some use cases. What is your idea?

File: src/System.Memory/ref/System.Memory.cs [hunk 2]
```diff
@@ -246,7 +246,7 @@ public static partial class SequenceReaderExtensions
         public static implicit operator System.Buffers.StandardFormat (char symbol) { throw null; }
         public static bool operator !=(System.Buffers.StandardFormat left, System.Buffers.StandardFormat right) { throw null; }
         public static System.Buffers.StandardFormat Parse(System.ReadOnlySpan<char> format) { throw null; }
-        public static System.Buffers.StandardFormat Parse(string format) { throw null; }
```
Comment (ahsonkhan) [inline src/System.Memory/ref/System.Memory.cs]: Keep this API as part of the ref since we shouldn't be removing it.

**Comment by ahsonkhan** [comment]
As an example, look at these two PRs that recently added a new API to CoreLib which was exposed from System.Runtime:
- https://github.com/dotnet/coreclr/pull/20773
- https://github.com/dotnet/corefx/pull/33225

And here is the auto-gen mirror PR:
- https://github.com/dotnet/corefx/pull/33245

The only difference here is that the API you are adding is being exposed from System.Memory.

**Comment by AlexejLiebenthal** [comment]
@ahsonkhan Thank you very much for your help and constructive feedback :) Now the things getting much clearer for me. 
I already thought that this has something to do with coreclr.
Hope I can finish my first PR to corefx tomorrow :)

**Comment by ahsonkhan** [comment]
> Now the things getting much clearer for me.
I already thought that this has something to do with coreclr.
Hope I can finish my first PR to corefx tomorrow :)

Sounds good. This doc provides more context that could be useful to you:

https://github.com/dotnet/coreclr/blob/master/Documentation/project-docs/adding_new_public_apis.md

File: src/Common/src/CoreLib/System/Buffers/StandardFormat.cs [hunk 3]
```diff
@@ -87,24 +109,20 @@ public static StandardFormat Parse(ReadOnlySpan<char> format)
                 {
                     uint digit = format[srcIndex] - 48u; // '0'
                     if (digit > 9)
-                        throw new FormatException(SR.Format(SR.Argument_CannotParsePrecision, MaxPrecision));
+                        return throws ? throw new FormatException(SR.Format(SR.Argument_CannotParsePrecision, MaxPrecision)) : false;
```
[the remark this answers is not in the document]
Comment (AlexejLiebenthal) [reply src/Common/src/CoreLib/System/Buffers/StandardFormat.cs]: Okay. For consistency shouldn't also 
[L55-58](https://github.com/dotnet/corefx/blob/6e86731f99ed0f6f35256c4ea4d64123f5b471c0/src/Common/src/CoreLib/System/Buffers/StandardFormat.cs#L55-L58) and [L96-97](https://github.com/dotnet/corefx/blob/6e86731f99ed0f6f35256c4ea4d64123f5b471c0/src/Common/src/CoreLib/System/Buffers/StandardFormat.cs#L96-97) be refactored witch brackets

[the remark this answers is not in the document]
Comment (ahsonkhan) [reply src/Common/src/CoreLib/System/Buffers/StandardFormat.cs]: No, neither of those require brackets. Based on the coding-guidelines and the pattern that is prevalent within the repo, there are exceptions for the following:
1) For argument validation checks that throw at the top of the method body.
2) Fir single line if statements at the root scope of a method (i.e. blocks that are not nested in another block).

From https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md
> 1. We use [Allman style](http://en.wikipedia.org/wiki/Indent_style#Allman_style) braces, where each brace begins on a new line. A single line statement block can go without braces but the block must be properly indented on its own line and must not be nested in other statement blocks that use braces (See issue [381](https://github.com/dotnet/corefx/issues/381) for examples). One exception is that a `using` statement is permitted to be nested within another `using` statement by starting on the following line at the same indentation level, even if the nested `using` contains a controlled block.

[the remark this answers is not in the document]
Comment (AlexejLiebenthal) [reply src/Common/src/CoreLib/System/Buffers/StandardFormat.cs]: Thank you again :+1:

**Comment by AlexejLiebenthal** [comment]
@dotnet-bot test

**Comment by AlexejLiebenthal** [comment]
@dotnet-bot test this please

**Comment by AlexejLiebenthal** [comment]
@dotnet-bot test Linux x64 Release Build
@dotnet-bot test NETFX x86 Release Build
@dotnet-bot test Packaging All Configurations x64 Debug Build
@dotnet-bot test UWP CoreCLR x64 Debug Build 
@dotnet-bot test Windows x64 Debug Build 
@dotnet-bot test Windows x86 Release Build

**Comment by ahsonkhan** [comment]
@dotnet-bot test this please

**Comment by AlexejLiebenthal** [comment]
@dotnet-bot test Linux x64 Release Build

@ahsonkhan Do you know what is going wrong here?
It seems like netfx build can't find the TryParse method/doesn't use the implementation of coreclr/System.Buffers:
``error CS0117: 'StandardFormat' does not contain a definition for 'TryParse'``

**Comment by ahsonkhan** [comment]
> Do you know what is going wrong here?
It seems like netfx build can't find the TryParse method/doesn't use the implementation of coreclr/System.Buffers

**Yes, this should be fixed. Here is the cause:**
- We used to build System.Memory (ref/src) for netstandard as well as netcoreapp:
  - https://github.com/dotnet/corefx/blob/448ba2875bdadf564fa2312a2b687ddbe9919ecb/src/System.Memory/src/Configurations.props#L4-L7
- We now only build for netcoreapp (no longer netstandard, since the OOB API surface is frozen):
  - https://github.com/dotnet/corefx/blob/3839c4a39ca2cb62f197b4bb92f777fc53b82038/src/System.Memory/src/Configurations.props#L4-L10
  - https://github.com/dotnet/corefx/pull/30897
  - https://github.com/dotnet/corefx/issues/30960
- The tests still run on netstandard as well, and expect an API that doesn't exist (for Netfx):
  - https://github.com/dotnet/corefx/blob/3839c4a39ca2cb62f197b4bb92f777fc53b82038/src/System.Memory/tests/Configurations.props#L6

**The fix:**
- Remove the netstandard configuration from the tests since the implementation doesn't build for it anymore.

If you are up for it, the fix shouldn't be too difficult. There might be some if/defs or netstandard specific test code you would have to remove. Let me know if you need any help or get blocked.

**Comment by AlexejLiebenthal** [comment]
Oh I was not so far away. I already opened this file corefx/src/System.Memory/src/Configurations.props earlier and played a bit around with no success with the build configuration but not the package configuration.
I will try to fix it till end of the week.

And thank you again :)

**Comment by AlexejLiebenthal** [comment]
@ahsonkhan  Stuck again :-/

After removing ``netstandard`` from ``corefx/src/System.Memory/tests/Configurations.props`` I now get a another build error (``C:\Program Files\dotnet\sdk\2.1.500\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.TargetFrameworkInference.targets(96,5): error NETSDK1013: The TargetFramework value '' was not recognized. It may be misspelled. If not, then the TargetFrameworkIdentifier and/or TargetFrameworkVersion properties must be specified explicitly.``) when I executed ``dotnet msbuild /p:TargetGroup=netfx /t:BuildAndTest``

Also with no success building with ``/p:ArchGroup=x86`` and some combinations of ``dotnet msbuild /p:TargetGroup=netfx /p:ArchGroup=x86 /p:TargetFramework=v4.7.2 /p:TargetFrameworkIdentifier=.NETFramework /t:BuildAndTest``

**Comment by ahsonkhan** [comment]
> when I executed ``dotnet msbuild /p:TargetGroup=netfx /t:BuildAndTest``

Once you have removed netstandard from `corefx/src/System.Memory/tests/Configurations.props`, you cannot run tests targeting netfx since the package/library (along with tests) no longer supports netfx.

So, it failing is expected, and isn't something that needs to be fixed. You can only pass in `TargetGroups` that are supported (i.e netcoreapp and uap).

**Comment by AlexejLiebenthal** [comment]
🤦‍♂️

**Pull request merged**